tools/PI/DevHome.PI/Helpers/SnapHelper.cs

```diff
@@ -0,0 +1,148 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.ComponentModel;
+using System.Diagnostics;
+using CommunityToolkit.Mvvm.ComponentModel;
+using DevHome.Common.Extensions;
+using DevHome.PI.Models;
+using DevHome.PI.ViewModels;
+using Microsoft.UI.Xaml;
+using Windows.Graphics;
+using Windows.Win32;
+using Windows.Win32.Foundation;
+using Windows.Win32.UI.Accessibility;
+using WinUIEx;
+
+namespace DevHome.PI.Helpers;
+
+public class SnapHelper
+{
+    // TODO The SnapOffsetHorizontal and UnsnapGap values don't allow for different DPIs.
+    private const int UnsnapGap = 9;
+
+    // It seems the way rounded corners are implemented means that the window is really 8px
+    // bigger than it seems, so we'll subtract this when we do sidecar snapping.
+    private const int SnapOffsetHorizontal = 8;
+
+    private readonly WINEVENTPROC _winPositionEventDelegate;
+    private readonly WINEVENTPROC _winFocusEventDelegate;
+
+    private HWINEVENTHOOK _positionEventHook;
+    private HWINEVENTHOOK _focusEventHook;
+
+    public SnapHelper()
+    {
+        _winPositionEventDelegate = new(WinPositionEventProc);
+        _winFocusEventDelegate = new(WinFocusEventProc);
+    }
+
+    public void Snap()
+    {
+        Debug.Assert(_positionEventHook == HWINEVENTHOOK.Null, "Hook should be cleared");
```

Comment:
This statement seems reversed.  I think that the event is cleared if it is null.  What does null signify in this case?

Reply:
No, it is not reversed. We want to make sure that the hook is null before we set them. Updated the message.